# Домашнє завдання 7

In [ ]:

!pip install surprise
from surprise import Dataset, accuracy
from surprise.model_selection import cross_validate, GridSearchCV
from surprise.prediction_algorithms import BaselineOnly, SVD, SVDpp, NMF

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.4/154.4 kB 3.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp310-cp310-linux_x86_64.whl size=2357228 sha256=70b4c3490cde8f74064040c9a47441069150d7ec6994c34735943554da447dd6
  Stored in directory: /root/.cache/pip/wheels/4b/3f/df/6acbf0a40397d9bf3ff97f582cc22fb9ce66adde75bc71fd54
Successfully built scikit-surprise


Завантажуємо movielens-100k датасет(алгоритм запропонує завантажити його, якщо необхідно)

In [ ]:
data = Dataset.load_builtin('ml-100k')

Dataset ml-100k could not be found. Do you want to download it? [Y/n] y
Trying to download dataset from https://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to /root/.surprise_data/ml-100k


Крос-валідація на моделі матричної факторизації SVD

In [ ]:
algorithm_svd = SVD()

cross_validate(algorithm_svd, data, measures=["RMSE", "MAE"], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9376  0.9345  0.9357  0.9361  0.9400  0.9368  0.0019  
MAE (testset)     0.7395  0.7376  0.7380  0.7377  0.7397  0.7385  0.0009  
Fit time          2.14    1.44    1.43    1.45    1.62    1.62    0.27    
Test time         0.23    0.21    0.12    0.21    0.13    0.18    0.05    


{'test_rmse': array([0.93764608, 0.93449393, 0.93569775, 0.93608374, 0.93996918]),
 'test_mae': array([0.73954806, 0.73755266, 0.73796633, 0.73769118, 0.73970396]),
 'fit_time': (2.144666910171509,
  1.4418690204620361,
  1.425452470779419,
  1.4544737339019775,
  1.6182899475097656),
 'test_time': (0.23493623733520508,
  0.20777201652526855,
  0.12427759170532227,
  0.20923328399658203,
  0.128922700881958)}

Результати крос-валідації для алгоритму SVD показують наступне:

* RMSE (середньо-квадратична помилка): Середнє значення: 0.9368 Стандартне відхилення: 0.0019

* MAE (середня абсолютна помилка): Середнє значення: 0.7385 Стандартне відхилення: 0.0009

* Час навчання (Fit time): Середнє значення: 1.62 сек Стандартне відхилення: 0.27 сек

* Час тестування (Test time): Середнє значення: 0.18 сек Стандартне відхилення: 0.05 сек

Ці результати вказують на те, що модель SVD має досить низьке значення середньо-квадратичної та середньої абсолютної помилок, що вказує на гарну точність моделі при прогнозуванні рейтингів. Також, час навчання та тестування є прийнятними і не сильно відрізняються між підсумковими фолдами.

Крос-валідація на моделі матричної факторизації SVD++

In [ ]:
algorithm_svdpp = SVDpp()

cross_validate(algorithm_svdpp, data, measures=["RMSE", "MAE"], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9201  0.9191  0.9141  0.9238  0.9207  0.9195  0.0032  
MAE (testset)     0.7195  0.7208  0.7155  0.7244  0.7230  0.7206  0.0031  
Fit time          27.80   27.40   27.23   27.94   27.98   27.67   0.30    
Test time         5.28    5.79    5.05    5.14    5.62    5.38    0.28    


{'test_rmse': array([0.92005996, 0.91907677, 0.91407886, 0.92379589, 0.92072172]),
 'test_mae': array([0.71945252, 0.7208028 , 0.71554574, 0.72441835, 0.72296175]),
 'fit_time': (27.800586938858032,
  27.39752960205078,
  27.233815670013428,
  27.944161891937256,
  27.982810735702515),
 'test_time': (5.282294034957886,
  5.788525819778442,
  5.0450663566589355,
  5.140308856964111,
  5.619417905807495)}

Результати крос-валідації для алгоритму SVDpp показують наступне:

* RMSE (середньо-квадратична помилка): Середнє значення: 0.9195 Стандартне відхилення: 0.0032

* MAE (середня абсолютна помилка): Середнє значення: 0.7206 Стандартне відхилення: 0.0031

* Час навчання (Fit time): Середнє значення: 27.67 сек Стандартне відхилення: 0.30 сек

* Час тестування (Test time): Середнє значення: 5.38 сек Стандартне відхилення: 0.28 сек

Алгоритм SVDpp показує гарні результати з точки зору точності прогнозів, яка вимірюється RMSE та MAE. Відзначимо, що SVDpp має більший час навчання та тестування порівняно з SVD, оскільки SVDpp включає додаткові параметри та фактори, що призводить до більшої обчислювальної складності. Важливо збалансувати точність та часові витрати в залежності від конкретних вимог нашого застосування.

Крос-валідація на моделі матричної факторизації NMF


In [ ]:
algorithm_nmf = NMF()

cross_validate(algorithm_nmf, data, measures=["RMSE", "MAE"], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9660  0.9593  0.9632  0.9640  0.9632  0.9632  0.0022  
MAE (testset)     0.7584  0.7559  0.7575  0.7554  0.7582  0.7571  0.0012  
Fit time          2.89    2.10    2.16    2.08    2.03    2.25    0.32    
Test time         0.33    0.11    0.23    0.10    0.19    0.19    0.08    


{'test_rmse': array([0.96601459, 0.95932937, 0.96321727, 0.96399744, 0.96324347]),
 'test_mae': array([0.75835023, 0.75593069, 0.75753141, 0.75541345, 0.75824676]),
 'fit_time': (2.8919243812561035,
  2.0994534492492676,
  2.160970449447632,
  2.0835752487182617,
  2.0319972038269043),
 'test_time': (0.3280637264251709,
  0.11219048500061035,
  0.22971487045288086,
  0.1001431941986084,
  0.1890249252319336)}

Результати крос-валідації для алгоритму NMF (Non-Negative Matrix Factorization) показують наступне:

* RMSE (середньо-квадратична помилка): Середнє значення: 0.9632 Стандартне відхилення: 0.0022

* MAE (середня абсолютна помилка): Середнє значення: 0.7571 Стандартне відхилення: 0.0012

* Час навчання (Fit time): Середнє значення: 2.25 сек Стандартне відхилення: 0.32 сек

* Час тестування (Test time): Середнє значення: 0.19 сек Стандартне відхилення: 0.08 сек

Алгоритм NMF також представляє певний рівень точності, але його значення RMSE та MAE трошки вищі порівняно з SVD та SVDpp. Важливо врахувати, що NMF використовує обмеження невід'ємності, що може бути корисним в деяких випадках, особливо якщо значення рейтингів не може бути від'ємним. Зате, час навчання та тестування для NMF є прийнятними.

Функція для підбору найкращих параметрів за допомогою крос-валідації


In [ ]:
def model_evaluation(algo, param_grid):
    print(f"\nAlgorithm: {algo.__name__}")

    # Створення об'єкта GridSearchCV
    grid_search = GridSearchCV(algo, param_grid, measures=['rmse', 'mae'], cv=3)

    # Запуск крос-валідації
    grid_search.fit(data)

    # Виведення найкращих параметрів та результатів
    print(f"Best parameters: {grid_search.best_params}")
    print(f"Best RMSE: {grid_search.best_score['rmse']:.4f}")

    # Оцінка моделі на весь набір даних (тренувальний та тестовий)
    algo_instance = grid_search.best_estimator['rmse']
    algo_instance.fit(data.build_full_trainset())
    predictions = algo_instance.test(data.build_full_trainset().build_testset())

    # Оцінка моделі
    print(f"\nTrainset Accuracy (RMSE): {accuracy.rmse(predictions):.4f}")

 Визначення найкращих параметрів моделі матричної факторизації SVD за допомогою крос-валідації

In [ ]:
param_grid = {
    'n_factors': [10, 20, 30],
    'n_epochs': [1, 5, 10],
    'lr_all': [0.002, 0.005],
    'reg_all': [0.02, 0.1]
}

algorithm_svd = SVD
model_evaluation(algorithm_svd, param_grid)


Algorithm: SVD
Best parameters: {'rmse': {'n_factors': 10, 'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.02}, 'mae': {'n_factors': 10, 'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.02}}
Best RMSE: 0.9491
RMSE: 0.9125

Trainset Accuracy (RMSE): 0.9125


Аналіз результатів:

* Після пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму SVD визначені як:

n_factors: 10 n_epochs: 10 lr_all: 0.005 reg_all: 0.02

* Найкращий RMSE: 0.9491

Найкращі параметри були використані для тренування моделі на повному тренувальному наборі даних, і її точність була оцінена на всьому наборі даних. Оцінка моделі на весь набір даних показала RMSE рівний 0.9125.

Загалом, результати показують, що модель SVD показала гарні характеристики. Найкращі параметри можуть бути змінені під час налаштування моделі. Значення RMSE є досить низькими, що свідчить про добру точність моделі у прогнозуванні рейтингів на всьому наборі даних. Таким чином, SVD із відповідними параметрами є ефективною моделлю для цього датасету, і вона може бути використана для рекомендаційного аналізу з задовільною точністю.

Визначення найкращих параметрів моделі матричної факторизації SVDpp за допомогою крос-валідації

In [13]:
param_grid_svdpp = {
    'n_factors': [10, 20, 30],
    'n_epochs': [1, 5, 10],
    'lr_all': [0.002, 0.005],
    'reg_all': [0.02, 0.1]
}

algorithm_svdpp = SVDpp
model_evaluation(algorithm_svdpp, param_grid_svdpp)


Algorithm: SVDpp
Best parameters: {'rmse': {'n_factors': 30, 'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.02}, 'mae': {'n_factors': 30, 'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.02}}
Best RMSE: 0.9408
RMSE: 0.8728

Trainset Accuracy (RMSE): 0.8728


 Аналіз результатів:

* Після пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму SVD визначені як:

n_factors: 30 n_epochs: 10 lr_all: 0.005 reg_all: 0.02

* Найкращий RMSE: 0.9408

Найкращі параметри були використані для тренування моделі на повному тренувальному наборі даних, і її точність була оцінена на всьому наборі даних. Оцінка моделі на весь набір даних показала RMSE рівний 0.8728.

Загалом, результати показують, що модель SVD++ показала гарні характеристики. Найкращі параметри можуть бути змінені під час налаштування моделі. Значення RMSE є досить низькими, що свідчить про добру точність моделі у прогнозуванні рейтингів на всьому наборі даних. Таким чином, SVD++ із відповідними параметрами також є ефективною моделлю для цього датасету, і вона може бути використана для рекомендаційного аналізу з задовільною точністю.

Визначення найкращих параметрів моделі матричної факторизації NMF за допомогою крос-валідації


In [ ]:
param_grid = {
    'n_factors': [10, 20, 30],
    'n_epochs': [1, 5, 10],
    'reg_pu': [0.06, 0.08, 0.1],
    'reg_qi': [0.06, 0.08, 0.1],
    'reg_bu': [0.02, 0.04, 0.6],
    'reg_bi': [0.02, 0.04, 0.6]
}
algorithm_nmf = NMF
model_evaluation(algorithm_nmf, param_grid)


Algorithm: NMF
Best parameters: {'rmse': {'n_factors': 10, 'n_epochs': 5, 'reg_pu': 0.1, 'reg_qi': 0.1, 'reg_bu': 0.02, 'reg_bi': 0.02}, 'mae': {'n_factors': 10, 'n_epochs': 5, 'reg_pu': 0.1, 'reg_qi': 0.1, 'reg_bu': 0.02, 'reg_bi': 0.02}}
Best RMSE: 1.0807
RMSE: 1.0396

Trainset Accuracy (RMSE): 1.0396


Аналіз результатів:

За результатами пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму NMF визначені як:
n_factors: 10 n_epochs: 5 reg_pu: 0.1 reg_qi: 0.1 reg_bu: 0.02 reg_bi: 0.02

Найкращий RMSE, отриманий за допомогою крос-валідації, становить 1.0807. Після отримання найкращих параметрів модель була навчена на повному тренувальному наборі даних, і її точність була оцінена на всьому наборі даних. Оцінка моделі на весь набір даних показала RMSE рівний 1.0396.

Висновки:

Отримані параметри та оцінки демонструють, що алгоритм NMF з обраною сіткою параметрів має гарні характеристики на даному наборі даних. Значення RMSE 1.0396 вказує на те, що середня квадратична помилка прогнозування рейтингів на весь набір даних досить низька, що може вказувати на ефективність моделі в прогнозуванні рейтингів. Загалом, отримані результати свідчать про те, що модель NMF з вибраними параметрами є ефективною для задачі рекомендаційного аналізу на даному датасеті.